examples/gno.land/r/gnoland/faucet/admin.gno

```diff
@@ -5,44 +5,77 @@ import (
 	"std"
 )
 
-func AdminSetInPause(inPause bool) error {
+func AdminSetInPause(inPause bool) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
 	gInPause = inPause
-	return nil
+	return ""
 }
 
-func AdminSetMessage(message string) error {
+func AdminSetMessage(message string) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
 	gMessage = message
-	return nil
+	return ""
 }
 
-func AdminSetPerTransferSend(send std.Coins) error {
+func AdminSetTransferLimit(amount int64) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
-	gPerTransferSend = send
-	return nil
+	gLimit = std.Coin{Denom: "ugnot", Amount: amount}
+	return ""
 }
 
-func AdminSetAdminAddr(addr std.Address) error {
+func AdminSetAdminAddr(addr std.Address) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
 	gAdminAddr = addr
-	return nil
+	return ""
 }
 
-func AdminSetControllerAddr(addr std.Address) error {
+func AdminAddController(addr std.Address) string {
 	if err := assertIsAdmin(); err != nil {
-		return err
+		return err.Error()
 	}
-	gControllerAddr = addr
-	return nil
+
+	size := gControllers.Size()
+
+	if size >= gControllersSize {
+		return "can not add more controllers than allowed"
+	}
+
+	if gControllers.Has(addr.String()) {
+		return addr.String() + " exists, no need to add."
+	}
+
+	gControllers.Set(addr.String(), addr)
+
+	return ""
+}
+
+func AdminRemoveController(addr std.Address) string {
+	if err := assertIsAdmin(); err != nil {
+		return err.Error()
+	}
+
+	if !gControllers.Has(addr.String()) {
+		return addr.String() + " is not on the controller list"
+	}
+
+	_, ok := gControllers.Remove(addr.String())
+
+	// it not should happen.
+	// we will check anyway to prevent issues in the underline implementation.
+
+	if !ok {
+		return addr.String() + " is not on the controller list"
+	}
```

Comment:
Do we need to error out here? Why not make this call noop, if the address wasn't in the controller list in the first place?

Reply:
We should check it and error out because the gController is a global variable and could be side load by other functions. Also, this is a public API. Whoever calls it does not need to know the underline implementation. They also want to know whether remove succeed or not and if that failed what is the reason.